File: app/src/main/java/com/fieldbook/tracker/traits/AngleTraitLayout.java
package com.fieldbook.tracker.traits;

import android.app.Activity;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.util.AttributeSet;
import android.widget.TextView;

import com.fieldbook.tracker.R;
import com.fieldbook.tracker.activities.CollectActivity;

//TODO this isn't showing in new trait creator
public class AngleTraitLayout extends BaseTraitLayout {
    SensorManager sensorManager;
    Sensor accelerometer;
    Sensor magnetometer;

    TextView pitchTv;
    TextView rollTv;
    TextView azimutTv;
    SensorEventListener mEventListener;

    public AngleTraitLayout(Context context) {
        super(context);
    }

    public AngleTraitLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AngleTraitLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public void setNaTraitsText() {
    }

    @Override
    public String type() {
        return "angle";
    }

    @Override
    public int layoutId() {
        return R.layout.trait_angle;
    }

    @Override
    public void init(Activity act) {
        sensorManager = (SensorManager) getContext().getSystemService(Context.SENSOR_SERVICE);
        accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        magnetometer = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD);

        pitchTv = act.findViewById(R.id.pitch);
        rollTv = act.findViewById(R.id.roll);
        azimutTv = act.findViewById(R.id.azimuth);

        mEventListener = new SensorEventListener() {
            float[] mGravity;
            float[] mGeomagnetic;
            Float azimut;
            Float pitch;
            Float roll;

            public void onAccuracyChanged(Sensor sensor, int accuracy) {
            }

            public void onSensorChanged(SensorEvent event) {
                if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER)
                    mGravity = event.values;
                if (event.sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD)
                    mGeomagnetic = event.values;
                if (mGravity != null && mGeomagnetic != null) {
                    float R[] = new float[9];
                    float I[] = new float[9];
                    boolean success = SensorManager.getRotationMatrix(R, I, mGravity, mGeomagnetic);
                    if (success) {
                        float orientation[] = new float[3];
                        SensorManager.getOrientation(R, orientation);
                        azimut = orientation[0]; // orientation contains: azimut, pitch and roll
                        pitch = orientation[1];
                        roll = orientation[2];

                        pitchTv.setText(Double.toString(Math.toDegrees(pitch)));
                        rollTv.setText(Double.toString(Math.toDegrees(roll)));
                        azimutTv.setText(Double.toString(Math.toDegrees(azimut)));
                    }
                }
            }
        };
    }

    @Override
    public void afterLoadNotExists(CollectActivity act) {
        super.afterLoadNotExists(act);
        sensorManager.registerListener(mEventListener, sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER),
                SensorManager.SENSOR_DELAY_NORMAL);
        sensorManager.registerListener(mEventListener, sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD),
                SensorManager.SENSOR_DELAY_NORMAL);
    }

    @Override
    public void deleteTraitListener() {
        ((CollectActivity) getContext()).removeTrait();
    }
}

File: app/src/main/java/com/fieldbook/tracker/traits/AngleTraitLayout.kt
package com.fieldbook.tracker.traits

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import com.fieldbook.tracker.R
import com.fieldbook.tracker.activities.CollectActivity
import com.fieldbook.tracker.utilities.GeodeticUtils.Companion.lowPassFilter
import com.fieldbook.tracker.views.CompassView
import com.google.android.material.floatingactionbutton.FloatingActionButton

/**
 * A trait for measuring the tilt angle using device orientation.
 * Uses the device tilt from gravity sensor (or accelerometer) for angle measurement.
 */
class AngleTraitLayout : BaseTraitLayout {

    companion object {
        private const val UPDATE_INTERVAL: Long = 100
    }

    private lateinit var compassView: CompassView
    private lateinit var captureButton: FloatingActionButton

    private var currentAngle: Float = 0f

    private val updateHandler: Handler = Handler(Looper.getMainLooper())
    private val updateRunnable = object : Runnable {
        override fun run() {
            updateFromSensor()
            updateHandler.postDelayed(this, UPDATE_INTERVAL)
        }
    }

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun deleteTraitListener() {
        clearObservationOrRemoveTrait()
        super.deleteTraitListener()
    }

    override fun setNaTraitsText() {}

    override fun type(): String = "angle"

    override fun layoutId(): Int = R.layout.trait_angle

    override fun init(act: Activity) {
        compassView = findTraitView(R.id.compassView)
        captureButton = findTraitView(R.id.captureButton)

        captureButton.setOnClickListener {
            val angleValue = "%.1f".format(currentAngle)
            updateObservation(currentTrait, angleValue)
            collectInputView.text = angleValue
        }

        // Sensor polling needs CollectController (device tilt). Skip outside Collect
        // (Constructor preview / node AndroidView hosts) — still show compass chrome.
        if (controller != null) {
            updateHandler.post(updateRunnable)
        }
    }

    private fun updateCompass() {
        compassView.setYawAngle(currentAngle)
    }

    private fun updateFromSensor() {
        val tilt = controller?.getDeviceTilt() ?: return
        currentAngle = lowPassFilter(floatArrayOf(tilt.roll), floatArrayOf(currentAngle))[0]
        updateCompass()
    }
}